% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/analyze_genome.R
\name{analyze_genome}
\alias{analyze_genome}
\alias{kmerize_fa}
\alias{find_kmer}
\alias{slide_genome}
\title{Sliding window and kmer exploration of genomes}
\usage{
kmerize_fa(path2fasta, kmerSize)

find_kmer(
  path2fasta,
  kmers,
  nCores = 1,
  revComp = TRUE,
  minSize2batch = 50,
  verbose = TRUE
)

slide_genome(path2fasta, bedObj, step, size, nCores = 1)
}
\arguments{
\item{path2fasta}{file.path to a fasta assembly file}

\item{kmerSize}{integer of length 1, specifying the size of the kmer to use}

\item{kmers}{character vector of kmers to search for}

\item{nCores}{integer of length 1, specifying the number of parallel
processes to run}

\item{revComp}{logical, should the kmers be reverse completemented too?}

\item{minSize2batch}{integer of length 1, specifying how many kmers should
be run in a single batch}

\item{verbose}{logical, should updates be printed to the console?}

\item{bedObj}{data.table or data.frame with chr, start, end columns}

\item{step}{integer of length 1, specifying the distance between adjacent
windows in bps.}

\item{size}{integer of length 1, specifying sliding window size in bps.}
}
\value{
All functions return a data.table, except kmerize_fasta, which
returns a character vector of kmers.
}
\description{
\code{analyze_genome} Sliding window and kmer exploration of genomes. These
functions are included here to assist with genome QC and comparative
analyses, but are not integrated into the GENESPACE pipeline.

\code{kmerize_fa} kmerize_fa

\code{find_kmer} find_kmer

\code{slide_genome} slide_genome
}
\details{
The kmer finding routines offer fast methods for exact string
matching, even for large sets of kmers or very large genome assemblies. The
sliding window method offers an alternative to froll from data.table,
which only can do fixed site width analyses. Here, we use base-pair windows,
which is usually what users want for genome comparisons.
}
\note{
\code{analyze_genome} is a generic name for the functions documented.
\cr
If called, \code{analyze_genome} returns its own arguments.
}
\examples{
\dontrun{
teloKmer <- c("CCCGAAA", "CCCTAAA")
teloLoc <- find_kmer(
  kmers = teloKmer,
  path2fasta = path2yourfasta)
swtelo <- slide_genome(
  path2fasta = path2yourfasta,
  windowSize = 5e3,
  stepSize = 500,
  bedObj = teloLoc)
ggplot(subset(swtelo$sw, grepl("LG", chr)), aes(x = Mbp, y = prop))+
  geom_line()+facet_wrap(~chr)
}

}
